### PR TITLE
Add a 'foreach' command

### PR DESCRIPTION
**Pull request by mmjconolly**
This is inspired by `git submodule foreach` [[link](https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-foreach--recursiveltcommandgt)]

```
$ wit foreach --help
usage: wit foreach [-h] [--quiet] [--continue-on-fail] cmd ...

Perform a command in each repository directory.
The repository list is created by reading records contained in 'wit-lock.json'.

If used, options --quiet and --continue-on-fail must be specified before the command.

The following environment variables are exposed to the command:
    WIT_REPO_NAME    repository name
    WIT_REPO_PATH    path to repository
    WIT_LOCK_SOURCE  inital source location of repository
    WIT_LOCK_COMMIT  commit recorded in lockfile, actual repository contents could be different
    WIT_WORKSPACE    path to root of wit workspace

positional arguments:
  cmd                 command to run in each repository
  args                arguments for the command

optional arguments:
  -h, --help          show this help message and exit
  --quiet             Supress printed lines indicating working repository
  --continue-on-fail  run the command in each repository regardless of
                      failures
```

```
$ wit foreach 'echo $WIT_REPO_NAME; git rev-parse HEAD'
Entering 'repo1'
repo1
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
Entering 'repo2'
repo2
bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
```

### CODE REVIEW

File: lib/wit/main.py [hunk 1]
```diff
@@ -108,6 +111,37 @@ def main() -> None:
         raise WitBug(e)
 
 
+def foreach(ws, args):
+    packages = sorted(ws.lock.packages, key=lambda x: str.lower(x.name))
```
Comment (richardxia) [inline lib/wit/main.py]:
```suggestion
    packages = sorted(ws.lock.packages, key=lambda x: x.name.lower())
```

I think it's a bit more idiomatic in Python to call this in method form rather than function form. See https://docs.python.org/3.5/library/stdtypes.html?highlight=lower#str.lower

File: lib/wit/main.py [hunk 2]
```diff
@@ -108,6 +111,37 @@ def main() -> None:
...
+
+    has_fail = False
+    for pkg in packages:
+        env = os.environ.copy()
+        env["WIT_REPO_NAME"] = pkg.name
+        env["WIT_REPO_PATH"] = str(ws.root / pkg.name)
```
Comment (richardxia) [inline lib/wit/main.py]: Is this the same as `pkg.repo.path`?

File: lib/wit/main.py (hunk 1 above)
Comment (jackkoenig) [inline lib/wit/main.py]: Rather than sorting, I think it should be left in the order read from the lock file. I thought that in `lock.py` we already lock the packages into an `OrderedDict`, but it looks like we only write them that way. My preference would be to make the ordering deterministic that way instead of via sorting here.

File: lib/wit/parser.py [hunk 3]
```diff
@@ -99,6 +99,33 @@ def err(msg):
...
+
+foreach_parser.add_argument('--quiet', action='store_true',
+                            help='Supress printed lines indicating working repository')
+foreach_parser.add_argument('--continue-on-fail', action='store_true',
+                            help='run the command in each repository regardless of failures')
+
+# 'cmd' and 'args' eventually become one list, but this forces at least one input string
+foreach_parser.add_argument('cmd', help='command to run in each repository')
+foreach_parser.add_argument('args', nargs=argparse.REMAINDER, help='arguments for the command')
+
```
Comment (jackkoenig) [inline lib/wit/parser.py]: Great job on the detailed description.

One minor clarification is that I was a little confused by what the environment variables mean when reading it. I would rephrase slightly to make it more clear that these are environment variables set by `wit` for each `command` rather than ones the user can set. I'm not sure on the best phrasing but here's a try:
```
Wit sets the following environment variables for each invocation of the command:
    WIT_REPO_NAME    repository name
    ...
```

File: lib/wit/main.py [hunk 4]
```diff
@@ -108,6 +111,37 @@ def main() -> None:
...
+        env["WIT_LOCK_SOURCE"] = pkg.source
+        env["WIT_LOCK_COMMIT"] = pkg.revision
+        env["WIT_WORKSPACE"] = str(ws.root)
+
+        if not args.quiet:
+            print("Entering '{}'".format(pkg.name))
```
Comment (jackkoenig) [inline lib/wit/main.py]: Should this be governed by the logger? I'm wondering if the whole `--quiet` option should be folded into that then this print would just be `log.info`.

File: lib/wit/main.py (hunk 2 above)
Comment (mmjconolly) [reply lib/wit/main.py]: Just tried it, it's value is `None`, I'm guessing we never create the repo object

Comment (jackkoenig) [reply lib/wit/main.py]: Yeah this is one of the weird parts of the implementation, I *think* `pkg.repo` is set by running `resolve` or something like that. We should probably set it for the `Package`s in the Lockfile though when we construct the `Workspace` object.

File: lib/wit/main.py (hunk 4 above)
Comment (mmjconolly) [reply lib/wit/main.py]: I'm trying to copy the usage from `git submodule foreach` usage, 
from the git docs:
```Unless given --quiet, foreach prints the name of each submodule before evaluating the command. ```

Comment (jackkoenig) [reply lib/wit/main.py]: I agree with generally following git's lead, but since we have unified logging (although I think it currently lacks a mechanism to _lower_ the verbosity which we obviously need), I would think it would be better to keep it unified with the logging API. I think it would be strange if just 1 command ignores your logging settings.

File: lib/wit/main.py (hunk 1 above)
Comment (mmjconolly) [reply lib/wit/main.py]: Happy to loop in the order they're in the lock file

File: lib/wit/main.py (hunk 4 above)
Comment (mmjconolly) [reply lib/wit/main.py]: If we make a global `--quiet`, we'd need to do it properly and make all the `print()` in the codebase conditional rather than just the log lines.

How about if I make the print line in this snippet unconditional, but print the line to stderr so that people who don't want it can filter it out with 2>/dev/null ?

Comment (mmjconolly) [reply lib/wit/main.py]: though grep-ing, we don't actually have that many `print()`

Comment (mmjconolly) [reply lib/wit/main.py]: I'll break this stuff into a separate diff

**Comment by mmjconolly** [comment]
Not sure how I closed this

File: lib/wit/main.py (hunk 4 above)
Comment (mmjconolly) [reply lib/wit/main.py]: https://github.com/sifive/wit/pull/214
